### DISCUSSION TITLE
eldoc displays signature instead of hover function

### DISCUSSION THREAD

**Discussion by Andre0991** [joaotavora/eglot#894] [Q&A] [answered]
Hi. I'm creating this as a discussion as I'm not sure what's the expected behaviour and if it's actually an issue.

The main issue is that the Clojure LSP server returns different information in the hover and signature actions, and `eldoc` ends up showing the one that is less useful.

### Steps to reproduce

1. Download clojure-lsp and add it to `PATH`:
```
mkdir repro-eglot
cd repro-eglot

# add to path
wget https://github.com/clojure-lsp/clojure-lsp/releases/download/2022.02.23-12.12.12/clojure-lsp
```
2. Clone this repo and a simple Clojure project:

```
wget https://raw.githubusercontent.com/jrblevin/markdown-mode/v2.5/markdown-mode.el
git clone https://github.com/joaotavora/eglot.git
git clone https://github.com/clojure/tools.trace.git
```

3. Start Emacs

```
emacs -Q -L eglot -l eglot.el -l markdown-mode.el tools.trace/src/main/clojure/clojure/tools/trace.clj -f eglot
```

`eglot` will ask for the server, use `clojure-lsp`.

4. Run `eldoc-doc-buffer` (just for reading the whole doc more confortably)

5. In any part of the file, start a new line and write `(assoc)`. Move the point to `a`.

The eldoc buffer will display a documentation that includes examples:

![image](https://user-images.githubusercontent.com/7083061/159490657-dc809e8e-c6c5-4d0a-87a9-979e09478fcd.png)

6. Now, move the point to `and` in the very last line of the file. See that the presented documentation does not include examples.

![image](https://user-images.githubusercontent.com/7083061/159490890-12835e9c-b1b7-4f2d-8281-3888b0f32fb1.png)

logs when the point is over *assoc*:

<details>

[client-request] (id:71) Tue Mar 22 10:24:24 2022:
(:jsonrpc "2.0" :id 71 :method "textDocument/signatureHelp" :params
	  (:textDocument
	   (:uri "file:///Users/andreperictavares/temp/repro-eglot/tools.trace/src/main/clojure/clojure/tools/trace.clj")
	   :position
	   (:line 437 :character 1)))
[client-request] (id:72) Tue Mar 22 10:24:24 2022:
(:jsonrpc "2.0" :id 72 :method "textDocument/hover" :params
	  (:textDocument
	   (:uri "file:///Users/andreperictavares/temp/repro-eglot/tools.trace/src/main/clojure/clojure/tools/trace.clj")
	   :position
	   (:line 437 :character 1)))
[client-request] (id:73) Tue Mar 22 10:24:24 2022:
(:jsonrpc "2.0" :id 73 :method "textDocument/documentHighlight" :params
	  (:textDocument
	   (:uri "file:///Users/andreperictavares/temp/repro-eglot/tools.trace/src/main/clojure/clojure/tools/trace.clj")
	   :position
	   (:line 437 :character 1)))
[server-reply] (id:73) Tue Mar 22 10:24:24 2022:
(:jsonrpc "2.0" :id 73 :result
	  [(:range
	    (:start
	     (:line 334 :character 26)
	     :end
	     (:line 334 :character 31)))
	   (:range
	    (:start
	     (:line 437 :character 1)
	     :end
	     (:line 437 :character 6)))])
[server-reply] (id:72) Tue Mar 22 10:24:24 2022:
(:jsonrpc "2.0" :id 72 :result
	  (:contents
	   (:kind "markdown" :value "```clojure\nclojure.core/assoc\n```\n\nassoc[iate]. When applied to a map, returns a new map of the\nsame (hashed/sorted) type, that contains the mapping of key(s) to\nval(s). When applied to a vector, returns a new vector that\ncontains val at index. Note - index must be <= (count vector).\n\n__Examples:__\n\n```clojure\n(assoc {} :key1 \"value\" :key2 \"another value\")\n;;=> {:key2 \"another value\", :key1 \"value\"}\n\n;; Here we see an overwrite by a second entry with the same key\n(assoc {:key1 \"old value1\" :key2 \"value2\"} \n        :key1 \"value1\" :key3 \"value3\")\n;;=> {:key3 \"value3\", :key2 \"value2\", :key1 \"value1\"}\n\n;; We see a nil being treated as an empty map\n(assoc nil :key1 4)\n;;=> {:key1 4}\n\n;; 'assoc' can be used on a vector (but not a list), in this way: \n;; (assoc vec index replacement)\n(assoc [1 2 3] 0 10)     ;;=> [10 2 3]\n(assoc [1 2 3] 2 '(4 6)) ;;=> [1 2 (4 6)]\n;; Next, 0 < index <= n, so the following will work\n(assoc [1 2 3] 3 10)     ;;=> [1 2 3 10]\n;; However, when index > n, an error is thrown\n(assoc [1 2 3] 4 10)\n;; java.lang.IndexOutOfBoundsException (NO_SOURCE_FILE:0)\n\n;; From http://clojure-examples.appspot.com/clojure.core/assoc\n```\n---\n```clojure\n;; here is an example of updating a field in a map.\n(def test-map {:account-no 12345678 :lname \"Jones\" :fnam \"Fred\"})\n(assoc test-map :fnam \"Sue\")\n;;=> {:account-no 12345678, :lname \"Jones\", :fnam \"Sue\"}\n\n;; notice that test-map is unchanged\ntest-map\n;;=> {:account-no 12345678 :lname \"Jones\" :fnam \"Fred\"})\n```\n---\n```clojure\n;; beware of this\n(assoc {} nil nil)\n;;=> {nil nil}\n```\n---\n```clojure\n;;transform a map´s values using reduce and assoc\n\n(defn transform\n  [coll]\n  (reduce (fn [ncoll [k v]]\n            (assoc ncoll k (* 10 v)))\n          {}\n          coll))\n\n(transform {:a 1 :b 2 :c 3})\n;;{:a 10 :b 20 :c 30}\n\n```\n---\n```clojure\n;;assoc applied to a vector\n\n(def my-vec [1 2 5 6 8 9])\n\n(assoc my-vec 0 77)\n;;[77 2 5 6 8 9]\n\n```\n---\n```clojure\n;; convert a vector into a set with assoc\n\n(def book-city  {:awards [\"Hugo\" \"World Fantasy Award\" \"Arthur C. Clarke Award\"\n            \"British Science Fiction Award\"]\n   :title \"The City and the City\"\n   :authors [{:birth-year 1972, :name \"China Miéville\"}]})\n\n(assoc book-city :authors (set (:authors book-city)))\n\n;; {:awards [\"Hugo\" \"World Fantasy Award\" \"Arthur C. Clarke Award\" \"British Science Fiction Award\"], :title \"The City and the City\", :authors #{{:birth-year 1972, :name \"China Miéville\"}}}\n```\n---\n```clojure\n;; You can also change multiple keys at once\n\n(def my-map {:name \"Victor\" :age 27 :city \"Barcelona\"})\n\n(assoc my-map :name \"Tom\" :age 47)\n\n;; {:name \"Tom\", :age 47, :city \"Barcelona\"}\n\n```\n---\n```clojure\n;; You can add some more fields (kvs) to your map\n(assoc {:type \"chicken\"} :home \"Kentucky\" :cooking-method \"fried\" :tasty \"definitely\")\n```\n---\n```clojure\n;; add new key-value from original key value\n(let [contact {:name \"kimim\" :home \"Hangzhou\"}\n      new-name \"ivy\"]\n  (assoc contact :modified? (not (= (contact :name) new-name)) :name new-name))\n;; {:name \"ivy\" :home \"Hangzhou\" :modified? true}\n\n```\n\n__See also:__\n\n[clojure.core/assoc-in](https://clojuredocs.org/clojure.core/assoc-in)\n\n[clojure.core/dissoc](https://clojuredocs.org/clojure.core/dissoc)\n\n[clojure.core/merge](https://clojuredocs.org/clojure.core/merge)\n\n[clojure.core/update](https://clojuredocs.org/clojure.core/update)\n\n[clojure.core/associative?](https://clojuredocs.org/clojure.core/associative?)\n\n__Notes:__\n\nHere is a version that will create a vector when the key is numerical.
\nThis may be useful instead of throwing an IndexOutOfBoundsException.
\n
\n<pre>
\n(defn assoc-in-idx [m [k & ks] v]
\n  (let [value (get m k (when (number? (first ks)) []))
\n	m (if (and (vector? m) (number? k) (-> m count (< k)))
\n	    (reduce (fn [m _] (conj m nil)) m (range (count m) k))
\n	    m)
\n	v (if ks
\n	    (assoc-in-idx value ks v)
\n	    v)]
\n    (assoc m k v)))
\n</pre>
\n\n---\nthe API is blurry When applied to a vector
\n
\n<pre>
\n;; should indicate following
\n(assoc vector index val) </pre>\n\n----\n\n*[/Users/andreperictavares/.m2/repository/org/clojure/clojure/1.9.0/clojure-1.9.0.jar:clojure/core.clj](zipfile:///Users/andreperictavares/.m2/repository/org/clojure/clojure/1.9.0/clojure-1.9.0.jar::clojure/core.clj)*")
	   :range
	   (:start
	    (:line 437 :character 1)
	    :end
	    (:line 437 :character 6))))
[server-reply] (id:71) Tue Mar 22 10:24:24 2022:
(:jsonrpc "2.0" :id 71 :result nil)

</details>

logs when the point is over *and*:

<details>

[client-request] (id:56) Tue Mar 22 10:20:55 2022:
(:jsonrpc "2.0" :id 56 :method "textDocument/signatureHelp" :params
	  (:textDocument
	   (:uri "file:///Users/andreperictavares/temp/repro-eglot/tools.trace/src/main/clojure/clojure/tools/trace.clj")
	   :position
	   (:line 434 :character 6)))
[client-request] (id:57) Tue Mar 22 10:20:55 2022:
(:jsonrpc "2.0" :id 57 :method "textDocument/hover" :params
	  (:textDocument
	   (:uri "file:///Users/andreperictavares/temp/repro-eglot/tools.trace/src/main/clojure/clojure/tools/trace.clj")
	   :position
	   (:line 434 :character 6)))
[client-request] (id:58) Tue Mar 22 10:20:55 2022:
(:jsonrpc "2.0" :id 58 :method "textDocument/documentHighlight" :params
	  (:textDocument
	   (:uri "file:///Users/andreperictavares/temp/repro-eglot/tools.trace/src/main/clojure/clojure/tools/trace.clj")
	   :position
	   (:line 434 :character 6)))
[server-reply] (id:58) Tue Mar 22 10:20:55 2022:
(:jsonrpc "2.0" :id 58 :result
	  [(:range
	    (:start
	     (:line 184 :character 7)
	     :end
	     (:line 184 :character 10)))
	   (:range
	    (:start
	     (:line 196 :character 7)
	     :end
	     (:line 196 :character 10)))
	   (:range
	    (:start
	     (:line 328 :character 12)
	     :end
	     (:line 328 :character 15)))
	   (:range
	    (:start
	     (:line 402 :character 46)
	     :end
	     (:line 402 :character 49)))
	   (:range
	    (:start
	     (:line 420 :character 46)
	     :end
	     (:line 420 :character 49)))
	   (:range
	    (:start
	     (:line 434 :character 5)
	     :end
	     (:line 434 :character 8)))])
[server-reply] (id:57) Tue Mar 22 10:20:55 2022:
(:jsonrpc "2.0" :id 57 :result
	  (:contents
	   (:kind "markdown" :value "```clojure\nclojure.core/and\n[]\n[x]\n[x & next]\n```\n\nEvaluates exprs one at a time, from left to right. If a form\nreturns logical false (nil or false), and returns that value and\ndoesn't evaluate any of the other expressions, otherwise it returns\nthe value of the last expr. (and) returns true.\n\n__Examples:__\n\n```clojure\nuser=> (and true true)\ntrue\n\nuser=> (and true false)\nfalse\n\nuser=> (and false false)\nfalse\n\nuser=> (and '() '())\n()\n\nuser=> (and '[] '[])\n[]\n\nuser=> (and 0 1)  ; Note that this is *not* bitwise 'and'\n1\n\nuser=> (and 1 0)\n0\n\n```\n---\n```clojure\n;; See examples for \"if\" explaining Clojure's idea of logical true\n;; and logical false.\n```\n---\n```clojure\n; Note that, and does not evaluate if the first value is false\nuser=> (and false nil)\nfalse\n\nuser=> (and nil false)\nnil\n\nuser=> (and false (println \"foo\"))\nfalse\n\nuser=> (and (println \"foo\") false)\nfoo\nnil\n\nuser=> (and nil nil)\nnil\n```\n---\n```clojure\n; From the Clojure 1.9 source code.\n(defn qualified-keyword?\n  \"Return true if x is a keyword with a namespace\"\n  [x] (and (keyword? x) (namespace x) true))\n\n; Note how the return value of and is value of the last expression.\nuser=> (qualified-keyword? :hi/there)\ntrue\n\n; If we instead define the function as:\n(defn qualified-keyword?\n  [x] (and (keyword? x) (namespace x)))\n; we get the namespace as return value:\nuser=> (qualified-keyword? :hi/there)\n\"hi\"\n```\n\n__See also:__\n\n[clojure.core/or](https://clojuredocs.org/clojure.core/or)\n\n[clojure.core/if](https://clojuredocs.org/clojure.core/if)\n\n[clojure.core/every-pred](https://clojuredocs.org/clojure.core/every-pred)\n\n__Notes:__\n\nNote <b>add</b> is a macro, so you cannot <b>apply</b> it. For example, there is a vector of some Boolean values <b>[true true false true]</b>, which you want to test to see if they are all true. The code below will not work: <pre>(apply add [true true false true]) ;won't work</pre>
\nInstead, use this: <pre>(every? identity [true  true false true])</pre> More discussion can be found at http://osdir.com/ml/clojure/2010-01/msg01242.html\n---\ntypo: 'and' instead of 'add'?\n\n----\n\n*[/Users/andreperictavares/.m2/repository/org/clojure/clojure/1.9.0/clojure-1.9.0.jar:clojure/core.clj](zipfile:///Users/andreperictavares/.m2/repository/org/clojure/clojure/1.9.0/clojure-1.9.0.jar::clojure/core.clj)*")
	   :range
	   (:start
	    (:line 434 :character 5)
	    :end
	    (:line 434 :character 8))))
[server-reply] (id:56) Tue Mar 22 10:20:55 2022:
(:jsonrpc "2.0" :id 56 :result
	  (:signatures
	   [(:label "(and [])" :documentation "Evaluates exprs one at a time, from left to right. If a form\n  returns logical false (nil or false), and returns that value and\n  doesn't evaluate any of the other expressions, otherwise it returns\n  the value of the last expr. (and) returns true." :parameters
		    [])
	    (:label "(and [x])" :documentation "Evaluates exprs one at a time, from left to right. If a form\n  returns logical false (nil or false), and returns that value and\n  doesn't evaluate any of the other expressions, otherwise it returns\n  the value of the last expr. (and) returns true." :parameters
		    [(:label "x")])
	    (:label "(and [x & next])" :documentation "Evaluates exprs one at a time, from left to right. If a form\n  returns logical false (nil or false), and returns that value and\n  doesn't evaluate any of the other expressions, otherwise it returns\n  the value of the last expr. (and) returns true." :parameters
		    [(:label "x")
		     (:label "& next")])]
	   :activeSignature 2 :activeParameter 0))
</details>

### Investigation

It seems that the issue is that the hover and signature methods get different values.

In fact, if I change those lines in `eglot--managed-mode`, I always get the complete documentation:

```
(eglot--setq-saving eldoc-documentation-functions
                        '(eglot-hover-eldoc-function
                          ;; eglot-signature-eldoc-function
			  ))
```

### Discussion

- If the hover information was ever present, I'd try to customize `eldoc`, because it takes too much space (if not using `eldoc-doc-buffer`). Ideally, I'd see the signature in `eldoc` in the echo area and I would ask for the complete documetation, if needed.
- A coworker said that Java, Dart and Js/Ts LSP servers do similar things (present more detailed documentation by the hover method). I don't use these servers so I don't know what's the exact difference, though.

**Comment by nemethf** [comment]
Have you tried customizing `eldoc-documentation-function` (singular)?  Eldoc can be customized in many ways: `M-x customize-group RET eldoc RET`.

**Reply by Andre0991** [reply]
Thanks. Changing it to `'eldoc-documentation-compose` displays the full documentation.

Just to confirm if I'm understanding how `eglot` defaults are meant to be used: `eldoc-documentation-functions` gets both hover and signature functions, and it's up to the user to customize `eldoc` if the result is not optimal. Does that make sense?
I'm just wondering if the defaults should be changed. But I don't have any proposal.

**Reply by Andre0991** [reply]
In any case, I'm wondering if there's any workflow that enables me to use `eldoc-documentation-compose` on demand.

The issue is that the echo area gets too big. I think I could set `eldoc-echo-area-use-multiline-p` to `nil`, but I'd rather customize the maximum number of lines. Is there such option? I haven't found it.

**Reply by Andre0991** [reply]
Never mind, just found it in the customization group. It's part of `Eldoc Echo Area Use Multiline P`, I can change it to 5 lines.

**Comment by Andre0991** [comment, marked as answer]
To sum it up, this config works well for my needs with `clojure-lsp`:

```
(setq eldoc-documentation-strategy 'eldoc-documentation-compose
      eldoc-echo-area-use-multiline-p 5)
```

**Reply by Andre0991** [reply]
Update: I don't know if something changed or it never worked globally, but I needed to wrap the code above in a hook provided by `eglot`, otherwise it would override `eldoc-documentation-strategy` in buffers managed by `eglot`.

```
(add-hook 'eglot-managed-mode-hook
	  ;; This displays full docs for clojure functions.
	  ;; See https://github.com/joaotavora/eglot/discussions/894
	  #'(lambda ()
	      (setq-local eldoc-documentation-strategy
			  #'eldoc-documentation-compose

			  eldoc-echo-area-use-multiline-p
			  5)))
```
